archive/unix_pack.ksy

```diff
@@ -0,0 +1,107 @@
+meta:
+  id: unix_pack
+  title: Unix pack(1)-compressed data
+  application:
+    - pack(1)
+    - unpack(1)
+    - pcat(1)
+  # The extension is specifically lowercase.
+  # Uppercase .Z is used for compress(1)-compressed data.
+  file-extension: z
+  xref:
+    justsolve: Pack
+    wikidata: Q11239298
+  license: CC0-1.0
+  endian: be
+doc: |
+  The Unix pack compression format, a very basic compression format based on
+  Huffman coding. The Unix `pack(1)` utility could compress files into
+  this format, and could be decompressed using the utilities `unpack(1)`
+  and `pcat(1)`.
+  
+  The pack format is long obsolete - it was superseded in the 1980s
+  by the LZW-based compression format of `compress(1)` (which is itself
+  almost never used anymore on modern Unix derivatives). Despite this,
+  the `gzip(1)` utility still supports decompressing (but not compressing)
+  data in the pack compression format.
+doc-ref:
+  - 'https://www.vidarholen.net/contents/blog/?p=691'
+  - 'https://github.com/koalaman/pack/blob/master/pack.hs'
+  - 'https://git.savannah.gnu.org/cgit/gzip.git/tree/unpack.c'
+  # The source code for the original Unix pack and unpack utilities
+  # (as of Research Unix, Edition 8) can be found in the archive v8.tar.bz2
+  # in the directory usr/src/cmd/pack.
+  - 'https://www.tuhs.org/Archive/Distributions/Research/Dan_Cross_v8/'
+seq:
+  - id: magic
+    contents: [0x1f, 0x1e]
+  - id: len_uncompressed
+    type: u4
+    doc: The length of the data when uncompressed, in bytes.
+  - id: tree
+    type: tree
+    doc: The Huffman tree used to compress the data.
+  - id: data
```

Comment:
I wonder if we can implement decoding right in KS.

Reply:
I don't think this is possible. To read the data, you need to iterate through the bits one by one and walk through the tree as indicated by each bit. This would require keeping state through multiple iterations of a `repeat`ed field, which isn't possible with Kaitai Struct. Even decoding the tree into a usable form would be difficult (or even impossible) because of how compactly it is stored in byte form.

For comparison, [here](https://github.com/dgelessus/ksf_stuff/blob/master/archive/unix_pack.py#L9-L66) is my Python code for decoding the tree and decompressing the data. If you have any idea how that could be translated to Kaitai Struct, feel free to try.